src/svlog/resolver.rs

```diff
@@ -737,16 +737,36 @@ impl<'a, 'c, C: Context<'a>> ScopeGenerator<'a, 'c, C> {
     }
 
     /// Register a definition.
-    pub fn add_def(&mut self, def: Def<'a>) {
+    pub fn add_def(&mut self, mut def: Def<'a>) {
         trace!(" - Adding definition {:?}", def);
 
         // Check that the definition does not collide with a previous one.
         if let Some(existing) = self.scope.defs.get(&def.name.value) {
-            // Do not redefine ports.
             match existing.node {
+                // Do not redefine ports.
                 DefNode::IntPort(_) => return,
-                _ => (),
+                // Handle being allowed to override forward declarations
+                DefNode::Ast(node) => {
+                    if let ast::AllNode::Typedef(ast) = node.as_all() {
+                        if let ast::ForwardType { kind: _ } = ast.ty.kind.data {
+                            def.may_override = true;
+                        }
+                    }
+                }
             };
+
+            // We can also have multiple forward declarations of the same name, must
+            // ignore all but the first one here. This also correctly handles the case
+            // where a "forward" declaration actually comes after the definition, which
+            // is apparently not an error in svlog.
+            if let DefNode::Ast(node) = def.node {
+                if let ast::AllNode::Typedef(ast) = node.as_all() {
+                    if let ast::ForwardType { kind: _ } = ast.ty.kind.data {
+                        return;
+                    }
+                }
+            }
+
```

Comment:
I like this approach, because it basically gets rid of forward declarations whenever an actual definition is encountered, such that *any* forward declaration being produced as the result of a name resolution indicates that no definition has been provided. Pretty neat.